cstore_reader.c

```diff
@@ -200,6 +214,85 @@ CStoreReadFooter(StringInfo tableFooterFilename)
 }
 
 
+/* DeserializeBlockData  deserializes requested data block for all columns. It
+ * uncompresses serialized data if necessary, sets deserialized buffer index at
+ * read state. The function also deallocates previous data buffer used to keep temporary
+ * uncompressed data buffer.
+ */
+static void
+DeserializeBlockData(TupleDesc tupleDescriptor, StripeData *stripeData,
+					 MemoryContext stripeReadContext, uint64 blockIndex,
+					 BlockDataBuffers **blockDataBuffersArray)
+{
+	int columnIndex = 0;
+	MemoryContext oldContext = MemoryContextSwitchTo(stripeReadContext);
+
+	for (columnIndex=0; columnIndex < stripeData->columnCount; columnIndex++)
+	{
+		ColumnData *columnData = stripeData->columnDataArray[columnIndex];
+
+		if (columnData != NULL)
+		{
+			ColumnBlockData *blockData = columnData->blockDataArray[blockIndex];
+			Form_pg_attribute *attributeFormArray = tupleDescriptor->attrs;
+			bool typeByValue = attributeFormArray[columnIndex]->attbyval;
+			int typeLength = attributeFormArray[columnIndex]->attlen;
+			char typeAlign = attributeFormArray[columnIndex]->attalign;
+			BlockDataBuffers * blockDataBuffers = blockDataBuffersArray[columnIndex];
+			StringInfo valueBuffer = NULL;
+
+			if (blockDataBuffers->serializedValueBuffer != NULL)
+			{
+				pfree(blockDataBuffers->serializedValueBuffer->data);
+				pfree(blockDataBuffers->serializedValueBuffer);
+				blockDataBuffers->serializedValueBuffer = NULL;
+			}
+
+			valueBuffer = DecompressBuffer(blockData->valueBuffer,
+										   blockData->valueCompressionType);
+
+			DeserializeBoolArray(blockData->existBuffer,
+								 blockDataBuffers->existsArray,
+								 blockData->rowCount);
+			DeserializeDatumArray(valueBuffer, blockDataBuffers->existsArray,
+								  blockData->rowCount, typeByValue, typeLength,
+								  typeAlign, blockDataBuffers->valueArray);
+
+			/* store newly allocated uncompressed data buffer to be freed later */
+			if (valueBuffer != blockData->valueBuffer)
+			{
+				blockDataBuffers->serializedValueBuffer = valueBuffer;
+			}
```

Comment:
For 2, (1) if it doesn't hurt the correctness of program, (2) and if freeing the buffer in uncompressed case makes the code a bit simpler (less if's, less functions, etc.), then we can do something like this:

```C
/* free previous block's data buffers */
pfree(blockDataBuffers->serializedValueBuffer->data);
pfree(blockDataBuffers->serializedValueBuffer);

/* decompress and deserialize current block's data */
valueBuffer = DecompressBuffer(blockData->valueBuffer,
							   blockData->valueCompressionType);
if (blockData->valueCompressionType != COMPRESSION_NONE)
{
	/* we don't need the original value buffer anymore */
	pfree(blockData->valueBuffer->data);
	pfree(blockData->valueBuffer);
}

DeserializeBoolArray(blockData->existBuffer,
					 blockDataBuffers->existsArray,
					 blockData->rowCount);
DeserializeDatumArray(valueBuffer, blockDataBuffers->existsArray,
					  blockData->rowCount, typeByValue, typeLength,
					  typeAlign, blockDataBuffers->valueArray);

/* save current block's data buffer to be freed when we're done with the block */
blockDataBuffers->serializedValueBuffer = valueBuffer;
```

We can even remove the ```if (blockData->valueCompressionType != COMPRESSION_NONE)``` block without hurting the correctness, but it may confuse the reader that we're freeing the original buffer for non-compressed, and not for compressed.

If we model the code as above, then we need to make some other changes (which is what I wanted to say in 3):

1. In ```CreateBlockDataBuffersArray()```, instead of initializing ```serializedValueBuffer``` to ```NULL```, we should initialize it to ```makeStringInfo()```. This is because in the above code we are unconditionally freeing ```serializedValueBuffer```.
2. We probably can remove ```ResetUncompressedDataBuffers```.

I am not sure if what I said plays well with memory contexts. Here I am trying to make the code a bit simpler. If you think this doesn't make the code simpler, please think if we can make the code simpler in any other way or not.

Thanks.